File: matiasmasso/Mat.NET/DTO/DTOProductRank.cs
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOProductRank
    {
        public Periods Period { get; set; }
        public List<Item> Items { get; set; }
        public List<DTOGuidNom> Zonas { get; set; }

        public List<DTOGuidNom> Brands { get; set; }

        public Units Unit { get; set; }

        public DTOLang Lang { get; set; }

        public enum Periods
        {
            Year,
            Quarter,
            Month
        }

        public enum Units
        {
            Units,
            Eur
        }

        public DTOProductRank(Periods oPeriod) : base()
        {
            Period = oPeriod;
            Items = new List<Item>();
        }

        public DTOYearMonth YearMonthFrom()
        {
            var retval = DTOYearMonth.Current();
            switch (Period)
            {
                case DTOProductRank.Periods.Year:
                    {
                        retval = DTOYearMonth.Previous(12);
                        break;
                    }

                case DTOProductRank.Periods.Quarter:
                    {
                        retval = DTOYearMonth.Previous(3);
                        break;
                    }

                case DTOProductRank.Periods.Month:
                    {
                        retval = DTOYearMonth.Previous();
                        break;
                    }
            }
            return retval;
        }


        public decimal Share(Item item)
        {
            decimal retval = 0;
            var tot = Items.Sum(x => x.Value);
            if (tot > 0)
                retval = item.Value / tot;
            return retval;
        }

        public class Item
        {
            public DTOProductCategory Product { get; set; }
            public decimal Value { get; set; }
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOProductRank.vb
Public Class DTOProductRank

    Property Period As Periods
    Property Items As List(Of Item)
    Property Zonas As List(Of DTOGuidNom)

    Property Brands As List(Of DTOGuidNom)

    Property Unit As Units

    Property Lang As DTOLang

    Public Enum Periods
        Year
        Quarter
        Month
    End Enum

    Public Enum Units
        Units
        Eur
    End Enum

    Public Sub New(oPeriod As Periods)
        MyBase.New
        _Period = oPeriod
        _Items = New List(Of Item)
    End Sub

    Public Function YearMonthFrom() As DTOYearMonth
        Dim retval = DTOYearMonth.current
        Select Case _Period
            Case DTOProductRank.Periods.Year
                retval = DTOYearMonth.Previous(12)
            Case DTOProductRank.Periods.Quarter
                retval = DTOYearMonth.Previous(3)
            Case DTOProductRank.Periods.Month
                retval = DTOYearMonth.Previous()
        End Select
        Return retval
    End Function


    Public Function share(item As Item) As Decimal
        Dim retval As Decimal = 0
        Dim tot = _Items.Sum(Function(x) x.Value)
        If tot > 0 Then
            retval = item.Value / tot
        End If
        Return retval
    End Function

    Public Class Item
        Property Product As DTOProductCategory
        Property Value As Decimal
    End Class
End Class
